### PR TITLE
Added auto-scale to the collector

### LINKED ISSUE #848: Auto-scale collector instances
Based on an (old) Gitter message from @black-adder, it would be good to let the Jaeger Operator automatically scale up/down Jaeger:

> As a rule of thumb, if you see that spans are being dropped, I'd go ahead and add more jaeger-collector hosts. If that doesn't mitigate, then scale out your cassandra cluster.

@objectiser then shared that:
> We should look to add some autonomic behaviour to the management of jaeger, by monitoring relevant metrics to understand when reporting/storing trace data is resulting in issues and take action (e.g. scale up).
>
> Possibly an action can initially try scaling up the collector, but if that does not change the no. of dropped spans significantly after a specified time, then it tries to scale up the storage.
> 
> Trickier situation would be how to determine when to scale down. That may be a combination of jaeger metrics with possibly some other factors (e.g. cpu utilisation, etc) - but only scaling down to a predefined minimum config.

With today's knowledge and tools, I think we can start with a simple approach of just using the Horizontal Pod Autoscaler (HPA) from Kubernetes, to scale up/down based on CPU and memory. The idea is that when there's a shortage of workers, the CPU will be close to the limit, and when the queues are full, the memory will be close to the limit as well (once jaegertracing/jaeger#943 is closed).

To me, the only piece of infra that should be scaled for now is the collector. The query isn't typically heavily used to the point of requiring dynamic scaling, and agent's are either scaled with the application (sidecar), or can't be scaled at all (daemon sets). The only other remaining component is the ingester, which could be dealt with in a second phase.

This leaves scaling of the storage out of the equation for now: we could either add them in a second phase, or delegate this action to the storage's operator.

### PR DESCRIPTION
**Pull request by jpkrohling**
A Horizontal Pod Autoscaler (HPA) was added in this PR, along with a new MinReplicas and MaxReplicas. With that, the collector should now automatically scale up and down based on the CPU and/or memory consumption. When none of the new properties are specified, the minimum amount of replicas is 1, while the maximum number of replicas is 100. The HPA configuration is added only when the deployment strategy is either production or streaming.

Closes #848, even though the scaling of the storage isn't implemented by this one.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
This shows how the auto-scaling works in OpenShift. Note that this should also work in plain Kubernetes, but I'm not able to generate enough load on my local machine with minikube + ES (1
GiB) + tracegen.

image

**Comment by jpkrohling** [comment]
@kevinearls not sure we want to add a new e2e test for this, but perhaps you might have a good idea that wouldn't be too fragile?

**Comment by jpkrohling** [comment]
I just ran a longer test this morning, showing that it scales up and down. First, I deployed the `simple-prod` instance, along with `tracegen` (10 replicas). After about 20 minutes, 10 replicas of collector were available. Removing the `tracegen` deployment caused the number of replicas to go down back to 1 after about 20 minutes. Then, I changed `simple-prod` to add a `maxReplicas` set to 5. Then, I deployed `tracegen` again, and verified that the collector gets scaled only up to 5 replicas. Removing the `tracegen` causes the collector to eventually settle at 1 replica again.

```
$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   1/1     1            1           2m2s
simple-prod-query       1/1     1            1           2m2s
tracegen                10/10   10           10          30s

$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   10/10   10           10          19m
simple-prod-query       1/1     1            1           19m
tracegen                10/10   10           10          17m

$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   1/1     1            1           44m
simple-prod-query       1/1     1            1           44m

$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   5/5     5            5           59m
simple-prod-query       1/1     1            1           59m
tracegen                10/10   10           10          11m

$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   5/5     5            5           71m
simple-prod-query       1/1     1            1           71m
tracegen                10/10   10           10          24m

$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   5/5     5            5           71m
simple-prod-query       1/1     1            1           71m

$ kubectl get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
simple-prod-collector   1/1     1            1           88m
simple-prod-query       1/1     1            1           88m
```

And here's a set of screenshots from OpenShift (older events are at the bottom):

image

image

image

File: deploy/crds/jaegertracing.io_jaegers_crd.yaml [hunk 1]
```diff
@@ -5310,12 +5310,26 @@ spec:
...
                     type: string
                   type: object
+                maxReplicas:
+                  description: MaxReplicas sets an upper bound to the autoscaling
+                    feature. When autoscaling is enabled and no value is provided,
+                    a default value is used.
+                  format: int32
+                  type: integer
+                minReplicas:
+                  description: MinReplicas sets a lower bound to the autoscaling feature.
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply deploy/crds/jaegertracing.io_jaegers_crd.yaml]: Not on our side. The HPA object has a default for this value (1):

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#horizontalpodautoscalerspec-v2beta2-autoscaling

File: deploy/examples/tracegen.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    matchLabels:
+      app: tracegen
+  template:
+    metadata:
+      labels:
+        app: tracegen
+    spec:
+      containers:
+      - name: tracegen
+        image: jaegertracing/jaeger-tracegen:latest
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply deploy/examples/tracegen.yaml]: Yes, opened an issue (#866) to track this. The first image will probably be 1.17 (next release).

File: pkg/deployment/collector.go [hunk 3]
```diff
@@ -190,6 +191,88 @@ func (c *Collector) Services() []*corev1.Service {
...
+		maxReplicas = int32(100) // we need to have an upper bound, and 100 seems like a "good" max value
+	}
+
+	labels := c.labels()
+	labels["app.kubernetes.io/component"] = "hpa-collector"
+	baseCommonSpec := v1.JaegerCommonSpec{
+		Labels: labels,
+	}
+
+	avgUtilization := int32(90)
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply pkg/deployment/collector.go]: Added a `defaultAvgUtilization` to hold the value, assigning it here to `avgUtilization`, as it's not possible to get pointers to constant values.

File: pkg/deployment/collector_test.go [hunk 4]
```diff
@@ -21,7 +21,7 @@ func init() {
 
 func TestNegativeReplicas(t *testing.T) {
 	size := int32(-1)
-	jaeger := v1.NewJaeger(types.NamespacedName{Name: "TestNegativeReplicas"})
+	jaeger := v1.NewJaeger(types.NamespacedName{Name: "my-instance"})
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply pkg/deployment/collector_test.go]: We started naming them all "my-instance" some time ago, as individual names don't bring much value and we had a few copy/paste mistakes.

**Comment by objectiser** [comment]
@jpkrohling Approval subject to tests passing :)

**Comment by jpkrohling** [comment]
Local run has shown that a new permission is missing. I'm testing it locally and will update the PR once I confirm the tests are passing.

**Pull request merged**